source/sigv4.c

```diff
@@ -794,6 +895,336 @@ static SigV4Status_t getCredentialScope( SigV4Parameters_t * pSigV4Params,
+                                  size_t index,
+                                  size_t valLen,
+                                  size_t trimmedLength )
+    {
+        bool ret = false;
+
+        /* Only trim spaces. */
+        if( isspace( value[ index ] ) )
+        {
+            /* The last character is a trailing space. */
+            if( ( index + 1 ) == valLen )
+            {
+                ret = true;
+            }
+            /* Trim if the next character is also a space. */
+            else if( isspace( value[ index + 1 ] ) )
+            {
+                ret = true;
+            }
+            /* It is a leading space if no characters have been written yet. */
+            else if( trimmedLength == 0U )
+            {
+                ret = true;
+            }
+        }
+
+        return ret;
+    }
+
+    static SigV4Status_t writeSignedHeaderToCanonicalRequest( size_t headerIndex,
+                                                              uint32_t flags,
+                                                              CanonicalContext_t * canonicalRequest )
+    {
+        char * pBufLoc;
+        size_t buffRemaining, keyLen = 0, i = 0, trimKeyLen = 0;
+        const char * headerKey;
+        SigV4Status_t sigV4Status = SigV4Success;
+
+        assert( canonicalRequest != NULL );
+        assert( canonicalRequest->pBufCur != NULL );
+
+        pBufLoc = canonicalRequest->pBufCur;
+        buffRemaining = canonicalRequest->bufRemaining;
+
+        keyLen = canonicalRequest->pHeadersLoc[ headerIndex ].key.dataLen;
+
+        headerKey = canonicalRequest->pHeadersLoc[ headerIndex ].key.pData;
+
+        for( i = 0; i < keyLen; i++ )
+        {
+            if( !( flags & SIGV4_HTTP_PATH_IS_CANONICAL_FLAG ) )
+            {
+                if( !( isTrimmableSpace( headerKey, i, keyLen, trimKeyLen ) ) )
+                {
+                    if( buffRemaining < 1 )
+                    {
+                        sigV4Status = SigV4InsufficientMemory;
+                        break;
+                    }
+                    else
+                    {
+                        *pBufLoc = tolower( headerKey[ i ] );
+                        pBufLoc++;
+                        trimKeyLen++;
+                        buffRemaining -= 1;
+                    }
+                }
+            }
+            else
+            {
+                *pBufLoc = canonicalRequest->pHeadersLoc[ headerIndex ].key.pData[ i ];
+                pBufLoc++;
+                buffRemaining -= 1;
+            }
+        }
+
+        if( buffRemaining < 1 )
+        {
+            sigV4Status = SigV4InsufficientMemory;
+        }
+        else
+        {
+            *pBufLoc = ';';
+            pBufLoc++;
+            /* Calculate the remaining buffer. 1 is added for ';' character. */
+            buffRemaining -= 1;
+            canonicalRequest->pBufCur = pBufLoc;
+            canonicalRequest->bufRemaining = buffRemaining;
+        }
```

Comment:
```suggestion
To avoid duplication of buffer copying code as well as add some helpful comments.
```suggestion
        bool shouldCopyChar = false;
        char charToCopy;
        size_t numOfCopiedBytes = 0;
        for( i = 0; i < keyLen; i++ )
        {
            shouldCopyChar = false;
            if( !( flags & SIGV4_HTTP_PATH_IS_CANONICAL_FLAG ) )
            {
                if( !( isTrimmableSpace( headerKey, i, keyLen, numOfCopiedBytes ) ) )
                {
                    charToCopy = tolower( headerKey[ i ] );
                    shouldCopyChar = true;
                }
            }
            else
            {
                charToCopy= canonicalRequest->pHeadersLoc[ headerIndex ].key.pData[ i ];
                shouldCopyChar = true;
            }
            
            /* Remaining buffer space should at least accommodate the character to copy and the trailing ";" */
            if( shouldCopyChar &&  ( buffRemaining <= 1 ) )
            {
                sigV4Status = SigV4InsufficientMemory;
            }
            else if( shouldCopyChar && ( buffRemaining > 1 ) ) 
            {
                *pBufLoc = charToCopy;
                pBufLoc++;
                buffRemaining =-1;
                numOfCopiedBytes;
            }
        }
        
        /* Space for ending ";" was accounted for while copying header field data to
          * canonical request buffer. */
        if( sigV4Status == SigV4Success )
        {
            assert( buffRemaining >= 1 );
            *pBufLoc = ';';
            pBufLoc++;
            /* Calculate the remaining buffer. 1 is added for ';' character. */
            buffRemaining -= 1;
            canonicalRequest->pBufCur = pBufLoc;
            canonicalRequest->bufRemaining = buffRemaining;
        }
```